### PR TITLE
Layer order now preserved on refresh or on load from URL or JSON server

### PR DESCRIPTION
**Pull request by manuel-castro**
This PR changes the "layers" property in the JSON state from an object to an array to preserve the layer order (see https://github.com/seung-lab/neuroglancer/issues/228).

### CODE REVIEW

**Comment by manuel-castro** [comment]
I signed it!

File: src/neuroglancer/layer_specification.ts [hunk 1]
```diff
@@ -127,10 +128,22 @@ export class TopLevelLayerListSpecification extends RefCounted implements LayerL
...
   restoreState(x: any) {
-    verifyObject(x);
     this.layerManager.clear();
-    for (let key of Object.keys(x)) {
-      this.layerManager.addManagedLayer(this.getLayer(key, x[key]));
+    if (Array.isArray(x)) {
+      // If array, layers have an order
+      for (const layerObj of x) {
+        verifyObject(layerObj);
+        const name = layerObj.name;
```
Comment (jbms) [inline src/neuroglancer/layer_specification.ts]:
```suggestion
        const name = verifyObjectProperty(layerObj, 'name', verifyString);
```

**Comment by jbms** [comment]
This looks good, but can you also fix the Python code to support this format as well?

**Comment by manuel-castro** [comment]
My mistake. Thanks for the feedback

**Comment by jbms** [comment]
Thanks --- ManagedLayer.to_json in viewer_state.py also needs to be fixed to include the name property now

Since there aren't unit tests for this stuff at the moment, please make sure to manually test that e.g. the examples in python/neuroglancer/examples still work.

**Comment by manuel-castro** [comment]
Sorry about that again, I had difficulty running some of the Python code, but I figured out what the problem was and was able to. I added the name property to ManagedLayer.to_json, removed the deletion of the name property in TypeScript for symmetry, and changed what I believe was a typo in one of the examples.

File: src/neuroglancer/layer_specification.ts [hunk 2]
```diff
@@ -127,10 +128,20 @@ export class TopLevelLayerListSpecification extends RefCounted implements LayerL
...
   restoreState(x: any) {
-    verifyObject(x);
     this.layerManager.clear();
-    for (let key of Object.keys(x)) {
-      this.layerManager.addManagedLayer(this.getLayer(key, x[key]));
+    if (Array.isArray(x)) {
+      // If array, layers have an order
+      for (const layerObj of x) {
+        verifyObject(layerObj);
+        const name = verifyObjectProperty(layerObj, 'name', verifyString);
```
Comment (jbms) [inline src/neuroglancer/layer_specification.ts]:
```suggestion
        const name = this.layerManager.getUniqueLayerName(verifyObjectProperty(layerObj, 'name', verifyString));
```

**Pull request merged**